// react-aria@3.49.0/dist/private/interactions/useFocusable.js
import { focusSafely as $f192c2f16961cbe0$export$80f3e147d781571c } from './focusSafely.js';
import { mergeProps as $bbaa08b3cd72f041$export$9d1611c77c2fe928 } from '../utils/mergeProps.js';
import { useFocus as $1e74c67db218ce67$export$f8168d8dd8fd66e6 } from './useFocus.js';
import { useKeyboard as $8296dad1a4c5e0dc$export$8f71654801c2f7cd } from './useKeyboard.js';
import { useSyncRef as $b7115c395c64f7b5$export$4debdb1a3f0fa79e } from '../utils/useSyncRef.js';
import React__default, { useRef, useEffect, useContext } from 'react';

/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */ 









let $d1116acdf220c2da$export$f9762fab77588ecb = /*#__PURE__*/ (React__default).createContext(null);
function $d1116acdf220c2da$var$useFocusableContext(ref) {
    let context = (useContext)($d1116acdf220c2da$export$f9762fab77588ecb) || {};
    ($b7115c395c64f7b5$export$4debdb1a3f0fa79e)(context, ref);
    // eslint-disable-next-line
    let { ref: _, ...otherProps } = context;
    return otherProps;
}
function $d1116acdf220c2da$export$4c014de7c8940b4c(props, domRef) {
    let { focusProps: focusProps } = ($1e74c67db218ce67$export$f8168d8dd8fd66e6)(props);
    let { keyboardProps: keyboardProps } = ($8296dad1a4c5e0dc$export$8f71654801c2f7cd)(props);
    let interactions = ($bbaa08b3cd72f041$export$9d1611c77c2fe928)(focusProps, keyboardProps);
    let domProps = $d1116acdf220c2da$var$useFocusableContext(domRef);
    let interactionProps = props.isDisabled ? {} : domProps;
    let autoFocusRef = (useRef)(props.autoFocus);
    (useEffect)(()=>{
        if (autoFocusRef.current && domRef.current) ($f192c2f16961cbe0$export$80f3e147d781571c)(domRef.current);
        autoFocusRef.current = false;
    }, [
        domRef
    ]);
    // Always set a tabIndex so that Safari allows focusing native buttons and inputs.
    let tabIndex = props.excludeFromTabOrder ? -1 : 0;
    if (props.isDisabled) tabIndex = undefined;
    return {
        focusableProps: ($bbaa08b3cd72f041$export$9d1611c77c2fe928)({
            ...interactions,
            tabIndex: tabIndex
        }, interactionProps)
    };
}

export { $d1116acdf220c2da$export$f9762fab77588ecb as FocusableContext, $d1116acdf220c2da$export$4c014de7c8940b4c as useFocusable };
